soc/SoC_OCV.py

```diff
@@ -35,47 +34,70 @@ def __init__(self):
 		self.soc = [x for (x, y) in self.points]
 		self.v_ocv =  [y for (x, y) in self.points]
 		# Graph of the error of the SOC over the entire discharge (% error)
-		# Create BatTestPlotLinear.py (for more accuracy)
 		
 	def get_soc_and_v_ocv(self):
 		return self.soc, self.v_ocv
+	
+	def get_points(self):
+		return self.points
+	
+	def get_cell_ocv(self, soc: float or int):
+		if (soc > self.soc[-1] or soc < self.soc[0]):
+			raise Exception("The soc value if out of predicted range")
+		
+		max_soc = len(self.soc) - 1
+		min_soc = 0
 
-	# 
-	def get_cell_ocv(self, soc2):
-		if (soc2 > self.soc[-1] or soc2 < self.soc[0]):
-			return "Value is out of predicted range", "Value is out of predicted range"
-		v_ocv1 = 0
-		v_ocv2 = 0
+		while (min_soc <= max_soc):
+			mid_soc = math.floor((max_soc + min_soc) / 2)
+			if (self.soc[mid_soc] < soc):
+				min_soc = mid_soc + 1
+			elif (self.soc[mid_soc] > soc):
+				max_soc = mid_soc - 1
+			else:
+				return self.v_ocv[mid_soc]
+		
+		min_soc = max_soc - 10
+		return self.__find_not_in_search__(soc, min_soc)
+	
+	def __find_not_in_search__(self, soc: int, min_soc: int):
+		ocv0 = 0
+		ocv1 = 0
+		soc0 = 0
 		soc1 = 0
-		soc3 = 0
-
-		for index in range(len(self.soc)):
-			if (self.soc[index] == soc2):
-				return self.v_ocv[index]
-			elif (self.soc[index] < soc2):
-				soc1 = self.soc[index]
-				v_ocv1 = self.v_ocv[index]
-			elif ((self.soc[index] > soc2)):
-				soc3 = self.soc[index]	
-				v_ocv2 = self.v_ocv[index]
+		for index in range(min_soc, len(self.soc)):
+			if (self.soc[index] < soc):
+				soc0 = self.soc[index]
+				ocv0 = self.v_ocv[index]
+			elif ((self.soc[index] > soc)):
+				soc1 = self.soc[index]	
+				ocv1 = self.v_ocv[index]
 				break
-		v_ocv3 = v_ocv2 - (((v_ocv2 - v_ocv1)*(soc3 - soc2)) / (soc3 - soc1))
-		return v_ocv3
+		ocv = round(((soc - soc0)*(ocv1 - ocv0) / (soc1 - soc0)) + ocv0, 6)
+		return ocv
+	
+	def plot_graph(self):
+		plt.plot(self.soc, self.v_ocv, c = 'r')
+		plt.title("SOC vs. OCV values based on test data")
+		plt.xlabel("State of Charge (SOC)")
+		plt.ylabel("Open Circuit Voltage (OCV)")
+		plt.grid(True)
+		plt.show()
 
 class test_SoC_OCV:
```

Comment:
Can we move this into a distinct file + test folder please 😃

Reply:
I'll make a folder inside the soc folder and save that graph there. Would you want me to save all the other graphs and non-python files in that folder as well?